[ENH] Added SurfStatResels

ReinderVosDeWael opened the pull request:
Finished SurfStatResels and corresponding tests. It _should_ now work for all cases.

sheyma commented:
I kinda solved the bug & made it much faster (all coming soon...)

but could you please tell me what is meant by "accumarray" here? did you forget to import it from a package?

https://github.com/MICA-MNI/BrainStat/blob/resels/surfstat/python/SurfStatResels.py#L256

ReinderVosDeWael commented:
Ow - yea thats an old matlab function that I removed because it was buggy and hard to fix. I thought I had removed all occurrences of it. 

See lines 311/312 on the correct replacement. I can push this for you or you can change it yourself - whatever is easier for you. 

Lines 311/312 (additional comments added):
#reselspvert = reselspvert + accumarray(v1, r3, size=np.array([v, 1])) # Old version
reselspvert += np.bincount(v1,r3,v) # New version.

sheyma commented:
I'll do that myself then, thanks for the prompt reply!

sheyma commented:
wait... there is another potential bug. I am working on it now.

ReinderVosDeWael commented:
Should I review the current changes or are you still working on something?

sheyma commented:
It would be great if you could review it, I am done on resels branch.
Thanks in advance!

ReinderVosDeWael commented:
As a matter of style, should we undo some of the added line breaks, especially for when similar lines of code occur several times? E.g. see line 296-306 of SurfStatResels. These lines are practically carbon copies of each other and having them on one line makes finding their differences really simple. The added line breaks makes the 'spot the difference' a bit more tedious.

ReinderVosDeWael commented:
Fixed a bug when slm.lat and slm.resl were provided and mask was all false. There still seems to be a bug occurring on about 33% of runs of test_8. Haven't been able to pinpoint what's causing it yet though.

ReinderVosDeWael commented:
Found a bug wherein not all three MATLAB output arguments were correctly parsed and fixed this - this caused a few tests to error again though. Also added an input argument to matlab_init_surfstat to allow for non-default surfstat paths.

ReinderVosDeWael commented:
There seems to be a bug in row_ismember that causes it to return nans. That's causing the second output arguments of the real tests to all have a mismatch with the MATLAB output.

ReinderVosDeWael commented:
All known bugs should now be fixed.

ReinderVosDeWael commented:
@sheyma let me know if you're OK to merge this and I'll resolve the conflict during the merge.

sheyma commented on surfstat/surfstat_wrap.py:
```diff
@@ -7,10 +7,10 @@
 import sys
 
-def matlab_init_surfstat():
+def matlab_init_surfstat(surfstat_path='./matlab'):
```
Aha, I've seen your intention using a user-defined matlab path here.
However, no test function needs this change. They all work fine with matlab_init_surfstat() line. So, I'd not change it, it can stay fixed as earlier.

ReinderVosDeWael replied:
The conflict arises because I changed the path but someone else changed it to output the engine. Whilst its indeed not strictly necessary for the tests - I've used matlab_init_surfstat in my own notebooks and these don't always have BrainStat/surfstat as their base directory.

ReinderVosDeWael replied:
Have resolved the conflict. The behavior without input arguments is exactly identical to before i.e. no tests should be affected, but matlab_init_surfstat (and its associated functions) can now also be used without enforcing a specific working directory.

sheyma replied:
cool.

sheyma commented:
> As a matter of style, should we undo some of the added line breaks, especially for when similar lines of code occur several times? E.g. see line 296-306 of SurfStatResels. These lines are practically carbon copies of each other and having them on one line makes finding their differences really simple. The added line breaks makes the 'spot the difference' a bit more tedious.

Yes, you are right. It is a styling issue. You can change the style as you want now.
In future, I'd like to establish code checking job using ´flake8´ as @OualidBenkarim  recommended.

sheyma commented:
Yo yo yo! 

`python -m pytest tests/test_SurfStatResels.py 
`
works like a machine! Great!
